sql/src/main/java/io/crate/execution/dsl/projection/ColumnIndexWriterProjection.java

```diff
@@ -115,9 +124,41 @@ public ColumnIndexWriterProjection(RelationName relationName,
             for (int i = 0; i < mapSize; i++) {
                 allTargetColumns.add(Reference.fromStream(in));
             }
+
+            int outputSize = in.readVInt();
+            if (outputSize > 0) {
+                var result = new ArrayList<Symbol>(outputSize);
+                for (int i = 0; i < outputSize; i++) {
+                    result.add(Symbols.fromStream(in));
+                }
+                outputs = result;
+            } else {
+                outputs = List.of();
+            }
+
+            int returnValueSize = in.readVInt();
+            if (returnValueSize > 0) {
+                returnValues = new ArrayList<>(returnValueSize);
+                for (int i = 0; i < returnValueSize; i++) {
+                    returnValues.add(Symbols.fromStream(in));
+                }
+            } else {
+                returnValues = List.of();
+            }
         } else {
+            returnValues = List.of();
+            outputs = List.of();
             allTargetColumns = List.of();
         }
+
+    }
+
+    public List<? extends Symbol> outputs() {
+        return outputs;
+    }
+
+    public List<Symbol> returnValues() {
```

Comment:
What is the difference between `returnValues` and `outputs`? Isn't it redundant to have both?

Reply:
They are not the same. `outputs` are the types of result e.g. long for rowcount, while `returnValues` are the expression of the `returning` clause evaluated against the document.

Reply:
But if `returnValues` are present, then the `outputs` must match. (Or at least have matching types). 

I think this needs at least some clarification via comments how they relate to each other. Or maybe utilize the outputs for both and have some flag that indicates if it is a result set or a single rowcount.

Reply:
I just tried your suggestion to use the `outputs` and a flag indicating what is the purpose of the `outputs` and it clearly does not make the code simpler, because now i have to treat `outputs` in different ways depending on the purpose. For now I added a comment with the relationship between output and returnvalues and will give it a second though also related to updates.